### PR TITLE
make it possible to use `Tensor` in factor graphs

### PR DESCRIPTION
**Pull request by marcrasi**
* Adds differentiable `init(flatTensor:)` and `.flatTensor` for converting fixed size vectors from/to `Tensor`, so that we can "escape" to `TensorFlow`-land in calculations that need `TensorFlow` functions.
* Adds `FixedShapeTensor` so that we can use `Tensor`s as variables and error vectors in factor graphs.

### CODE REVIEW

**Comment by dellaert** [comment]
@marcrasi, I'm so glad to have you aboard :-) This would have taken us ages to figure out!

**Comment by marcrasi** [comment]
> @marcrasi, I'm so glad to have you aboard :-) This would have taken us ages to figure out!

I'm glad to help :)

**Comment by marcrasi** [comment]
I've updated this with more docs and explanation so it should be ready for review & merge now.

File: Sources/SwiftFusion/Core/FixedShapeTensor.swift [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import TensorFlow
```
Comment (dellaert) [inline Sources/SwiftFusion/Core/FixedShapeTensor.swift:1]: Maybe provide some docs and rationale for why this module exists.

File: Sources/SwiftFusion/Core/FixedShapeTensor.swift [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    var scalars = self.tensor.scalars
+    let r = try scalars.withUnsafeMutableBufferPointer { b in
+      try body(b)
+    }
+    self.tensor = Tensor(shape: Self.shape, scalars: scalars)
+    return r
+  }
+}
+
+public struct Tensor10_10: AdditiveArithmetic, FixedShapeTensor {
```
Comment (dellaert) [inline Sources/SwiftFusion/Core/FixedShapeTensor.swift:67]: a bit weird to have just 10,10 here. Is that for a test?

Comment (ProfFan) [reply Sources/SwiftFusion/Core/FixedShapeTensor.swift:67]: It's for my PPCA, this is still very experimental, so I'll merge now and improve later.

**Comment by marcrasi** [comment]
> where is flatTensor defined? Seems to suddenly pop up int his PR, but maybe I missed it

It's defined in this PR, in EuclideanVectorN.swift

File: Sources/SwiftFusion/Core/FixedShapeTensor.swift (hunk 2 above)
Comment (marcrasi) [reply Sources/SwiftFusion/Core/FixedShapeTensor.swift:67]: currently `EuclideanVectorN` (which is used in all the factor graph infra) has a requirement that the size is known statically at compiletime. the easiest way to do this is to define separate types for all the shapes that we want to use. I have some simplifications in mind that might let us lift the restriction that the size be known at compiletime, so I'll try to remove this weird type by implementing those.

File: Sources/SwiftFusion/Core/FixedShapeTensor.swift (hunk 1 above)
Comment (marcrasi) [reply Sources/SwiftFusion/Core/FixedShapeTensor.swift:1]: Ok, I'll do this in a follow up PR!

**Comment by marcrasi** [comment]
Oopsie, I have just realized that I never actually pushed the commit with my doc comments before claiming that this was commented and ready for review! Sorry about that. I'll open a new PR with that commit.

**Pull request merged**